semantic-core/src/Data/Core/Parser.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE ExplicitForAll, FlexibleInstances, LambdaCase, MultiParamTypeClasses, OverloadedLists,
+             ScopedTypeVariables #-}
+
+module Data.Core.Parser
+  ( module Text.Trifecta
+  , core
+  , lit
+  , expr
+  , lvalue
+  ) where
+
+-- Consult @doc/grammar.md@ for an EBNF grammar.
+
+import           Control.Applicative
+import qualified Data.Char as Char
+import           Data.Core
+import           Data.Name
+import           Data.Semigroup
+import           Data.String
+import qualified Text.Parser.Token as Token
+import qualified Text.Parser.Token.Highlight as Highlight
+import           Text.Trifecta hiding (ident)
+
+-- * Identifier styles and derived parsers
+
+validIdentifierStart :: Char -> Bool
+validIdentifierStart c = not (Char.isDigit c) && isSimpleCharacter c
+
+coreIdents :: TokenParsing m => IdentifierStyle m
+coreIdents = Token.IdentifierStyle
+  { _styleName              = "core"
+  , _styleStart             = satisfy validIdentifierStart
+  , _styleLetter            = satisfy isSimpleCharacter
+  , _styleReserved          = reservedNames
+  , _styleHighlight         = Highlight.Identifier
+  , _styleReservedHighlight = Highlight.ReservedIdentifier
+  }
+
+reserved :: (TokenParsing m, Monad m) => String -> m ()
+reserved = Token.reserve coreIdents
+
+identifier :: (TokenParsing m, Monad m, IsString s) => m s
+identifier = choice [quote, plain] <?> "identifier" where
+  plain = Token.ident coreIdents
+  quote = between (string "#{") (symbol "}") (fromString <$> some (noneOf "{}"))
+
+-- * Parsers (corresponding to EBNF)
+
+core :: (TokenParsing m, Monad m) => m Core
+core = expr
+
+expr :: (TokenParsing m, Monad m) => m Core
+expr = chainl1 atom go where
```

Comment:
I have an extremely mild preference for writing chains like this infix:

```haskell
expr = atom `chainl1` choice [ … ]
```

I definitely would not hold up merge for this, just wanted to point it out as the idiom isn’t always obvious but might be handy in future.